tls/s2n_handshake_io.c

```diff
@@ -635,6 +636,42 @@ static int handshake_read_io(struct s2n_connection *conn)
     return 0;
 }
 
+static void s2n_write_keylog(struct s2n_connection *conn)
+{
+    char *keylog_fn;
+    FILE *keylog_fp;
+
+    keylog_fn = getenv("SSLKEYLOGFILE");
+    if (!keylog_fn) {
+        return;
+    }
+
+    keylog_fp = fopen(keylog_fn, "a");
```

Comment:
We use GUARD() to check every call and return value. A lot of these functions can fail, like the fprintfs and so on ... and could result in nonsense output. In general it's be better to have an int type function and to propagate the errors.  Though here we need to make to close the fd in the error cases too. 

Rather than using fprintf/fopen/fclose directly would it be better to add these as stuffer abstractions? E.g. make a file writable stuffer? We have a file readable stuffer already, so that would be symmetrical. 

Typically too programs that write to appended log files also need to do things like handle the HUP signal. Are we going to need that kind of infrastructure in programs? Would we better off having a callback instead and letting the caller do the logging?

Reply:
> We use GUARD() to check every call and return value

But `GUARD` assumes a classic C API returning a negative number in case of failures. `fopen` returns `NULL`.

> Rather than using fprintf/fopen/fclose directly would it be better to add these as 
> stuffer abstractions? E.g. make a file writable stuffer? We have a file readable 
> stuffer already, so that would be symmetrical.

`s2n_stuffer_send_to_fd` already exists, and my first idea was to use it, in particular since there's `s2n_stuffer_hex` already - but it's part of the testlib, that's why I eventually didn't.

> Typically too programs that write to appended log files also need to do things like
>  handle the HUP signal. Are we going to need that kind of infrastructure in 
> programs? Would we better off having a callback instead and letting the caller do 
> the logging?

Good point. What I like about this is that it'd more in sync with the SSL_CTX_set_keylog_callback concept used by others. I'll update the PR.